rust/kernel/chrdev.rs

```diff
@@ -20,10 +19,57 @@ use crate::error::{Error, KernelResult};
 use crate::file_operations;
 use crate::types::CStr;
 
+/// Character device.
+///
+/// # Invariants
+///
+/// The pointer [`Cdev::cdev`] is valid and non-null.
+struct Cdev(*mut bindings::cdev);
+
+impl Cdev {
+    fn alloc(
+        fops: &'static bindings::file_operations,
+        module: &'static crate::ThisModule,
+    ) -> KernelResult<Self> {
+        // SAFETY: `bindings::cdev_alloc` returns a valid pointer or null.
+        let cdev = unsafe { bindings::cdev_alloc() };
+        if cdev.is_null() {
+            return Err(Error::ENOMEM);
+        }
+        // SAFETY: [`Cdev::cdev`] is valid and non-null.
+        // `fops` and `module` have `'static` lifetime.
+        unsafe {
+            (*cdev).ops = fops;
+            (*cdev).owner = module.0;
+        }
+        // INVARIANTS: [`Cdev::cdev`] is guaranteed valid and non-null.
+        Ok(Self(cdev))
+    }
+
+    fn add(&mut self, dev: bindings::dev_t, count: c_types::c_uint) -> KernelResult<()> {
+        // SAFETY: [`Cdev::cdev`] is valid and non-null.
+        // dev and count are scalars.
+        let rc = unsafe { bindings::cdev_add(self.0, dev, count) };
+        if rc != 0 {
+            return Err(Error::from_kernel_errno(rc));
+        }
+        Ok(())
+    }
+}
+
+impl Drop for Cdev {
+    fn drop(&mut self) {
+        // SAFETY: [`Cdev::cdev`] is guaranteed valid and non-null.
```

Comment:
Thanks so much for the review @ojeda ! I will fix up those `SAFETY` comments. My internal safety comment generator isn't properly calibrated yet, bear with me :)

Reply:
No worries at all -- thanks to you for being one of the early ones joining. We all needed to get accustomed, and if this effort is successful, we will have many other kernel developers asking the same things sooner or later. So hopefully these reviews help other readers as well (we can even reuse a few good reviews as documentation for newcomers).

Reply:
It's awesome being a guinea pig :)

Reply:
Just to be clear, we should avoid linking to the PRs themselves unless everyone involved agrees and instead write better docs based on them :-)

(The reason is that some people may prefer not to be shown in that light, i.e. asking "newbie" questions -- which they are not -- we all asked the same questions at some point)

Reply:
I find this experience extremely interesting - it shows how much `SAFETY` rigour is/will be required to guarantee a safe interface between kernel C and Rust drivers. One spends a few orders of magnitude more effort in the kernel Rust core, in order to get Rust driver safety.

A few potential problems spring to mind, though:
- human fallibility regarding initial `SAFETY` reasoning
- even more importantly, kernel C behaviour changing due to refactoring or bug fixes, and existing `SAFETY` reasoning going stale. this could be a serious concern, as the kernel is in constant flux (internally).

On the whole I believe Rust-for-Linux has unbelievable potential (that's why I'm here!). The optics and politics of the above could be an issue. The more technical, automated mitigations we can find, the better.